### PR TITLE
feat: add method to initialize SDK with an httpClient

### PR DESCRIPTION
**Pull request by jpill**
- Adds the ability to initialize SDK with an HttpClient, and to make modifications to subsequent requests
- Adds cancellation token to client
- Adds HttpResponseMessage to ShipEngineException
- Adds missing markdown
- Redefines several properties as nullable reference types
- Adds and implements IShipEngine interface with virtual mockable ShipEngine

### CODE REVIEW

File: ShipEngine/ShipEngine.cs [hunk 1]
```diff
@@ -1,16 +1,34 @@
...
 using System.Net.Http;
 using System.Text.Json;
-using System.Text.Json.Serialization;
 using System.Threading.Tasks;
 
 namespace ShipEngineSDK
 {
+    public static class ShipEngineExtensions
+    {
+        public static IHostApplicationBuilder AddShipEngine(this IHostApplicationBuilder builder, Action<HttpClient>? clientOverride = null)
```
Comment (joshuaflanagan) [inline ShipEngine/ShipEngine.cs:14]:
```suggestion
        public static IHostApplicationBuilder AddShipEngine(this IHostApplicationBuilder builder, Action<HttpClient>? configureClient = null)
```

File: ShipEngine/ShipEngine.csproj [hunk 2]
```diff
@@ -28,9 +28,12 @@
 			<PrivateAssets>all</PrivateAssets>
 			<IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
 		</PackageReference>
+		<PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="8.0.0" />
```
Comment (joshuaflanagan) [inline ShipEngine/ShipEngine.csproj:31]: Do these all work with .NET Framework 4.8? Or are they properly ignored?

File: ShipEngine/ShipEngineClient.cs [hunk 3]
```diff
@@ -21,28 +22,30 @@ public class ShipEngineClient
...
-            JsonSerializerOptions = new JsonSerializerOptions
-            {
-                DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
-                PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
-                PropertyNameCaseInsensitive = true,
-                WriteIndented = true,
-                Converters = { new JsonStringEnumMemberConverter() }
-            };
-        }
+        protected Action<HttpRequestMessage>? ModifyRequest { get; set; }
```
Comment (joshuaflanagan) [inline ShipEngine/ShipEngineClient.cs:39]: Does this need to be protected?  We want users to be able to set it. I know we talked about a `.WithModifiedRequest()` method that would return a new instance with this field set, but is there any reason to force them to use that method? Seems like it should be fine to let people do `client.ModifyRequest = r => ...` as well

File: ShipEngine/ShipEngineClient.cs [hunk 4]
```diff
@@ -71,6 +74,39 @@ public static HttpClient ConfigureHttpClient(Config config, HttpClient client)
...
+        /// <param name="client">The HttpClient to be configured</param>
+        /// <param name="apiKey">The API key to be used for all ShipEngine API calls</param>
+        /// <param name="baseUri">The base URI for the ShipEngine API</param>
+        /// <param name="timeout">The timeout for the ShipEngine API Calls</param>
+        /// <returns></returns>
+        public static HttpClient ConfigureHttpClient(HttpClient client, string apiKey, Uri? baseUri, TimeSpan? timeout = null)
+        {
+            client.DefaultRequestHeaders.Accept.Clear();
+
+            var osPlatform = Environment.OSVersion.Platform.ToString();
```
Comment (joshuaflanagan) [inline ShipEngine/ShipEngineClient.cs:90]: The duplication of all of this logic seems wrong. Can the non-static version invoke the static version?

Comment (jpill) [reply ShipEngine/ShipEngineClient.cs:90]: good callout, I have modified this in the updated PR here: https://github.com/ShipEngine/shipengine-dotnet/pull/93

File: ShipEngine/ShipEngineClient.cs (hunk 3 above)
Comment (jpill) [reply ShipEngine/ShipEngineClient.cs:39]: fair question - I think likely not, but am just going to remove the modify request bit from this effort while that gets further consideration, so as not to cause more delay with the rest of the functionality I'm looking to deliver here

**Comment by jpill** [comment]
closing this PR in favor of https://github.com/ShipEngine/shipengine-dotnet/pull/93

**Pull request closed without merging**